[quant][graphmode] Add quantized conv2d-relu fusion pattern

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33650 [not4land] debugging
* #33533 [not4land] Temporary workaround to get quantized ops
* #34572 [quant][graphmode] quantization support for aten::add
* #34518 [quant][graphmode] quantization work for prim::If
* #34571 [quant][graphmode][fix] use observed_values_ to check values are observed
* #34414 [quant][graphmode][refactor] insertObservers for Block
* #34411 [quant][graphmode][fix] Insert dequantize before use node
* #34349 [quant][graphmode][fix] preserve the type of original value when inserting dequant node
* #34348 [quant] Add dequantize.tensors
* #34347 [quant][graphmode] Add quantization support for aten::dropout
* #34346 [quant][graphmode] Add quantization support for aten::cat
* #34345 [quant][graphmode] Add prim::ListConstruct to general op handling
* #34315 [quant][graphmode] insert quant/dequant work for duplicated debugName
* #34314 [quant][graphmode][refactor] Change QParamMap to QParamVector
* #33854 [quant][graphmode] Quantization pattern for aten::linear
* #33532 [quant][graphmode] Add quantization pattern for quantized::add_relu
* **#33279 [quant][graphmode] Add quantized conv2d-relu fusion pattern**
* #33927 [quant][graphmode] Add Finalize function that inlines graph and produce quantized ops

Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20354878](https://our.internmc.facebook.com/intern/diff/D20354878)

raghuramank100 commented on torch/csrc/jit/passes/quantization_patterns.h:
```diff
@@ -63,8 +63,35 @@ graph(%packed_params, %a_quant, %r_scale, %r_zero_point, %r_dtype):
         %r = quantized::linear(%a_quant, %packed_params, %r_scale, %r_zero_point)
         return (%r) )";
 
+    std::string conv2d_relu = R"(
```
I assume that this pattern is after inlining as I dont see module call methods here.

jerryzh168 replied:
yes, this is quant fusion pattern

raghuramank100 commented on test/test_jit.py:
```diff
@@ -1500,6 +1502,28 @@ def test_quant_fusion(self):
+        m(data)
+        m = convert_script(m, True)
+        print(m.graph_for(data))
+        FileCheck().check_not("aten::conv2d") \
```
Can you run this test? How does making attributes constant and inlining occur?

jerryzh168 replied:
this is some old test, this is just testing graph rewrite, probably not that useful

jerryzh168 commented on test/test_jit.py:
```diff
@@ -1480,6 +1482,28 @@ def test_quant_fusion(self):
             torch._C._jit_pass_quant_fusion(graph)
             FileCheck().run(input_str, graph)
 
+    def test_quantized_conv_relu_fusion(self):
```
@raghuramank100 the new test is going to test the fusion since we do inline etc. in convert script

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

raghuramank100 commented on test/test_jit.py:
```diff
@@ -1568,6 +1571,27 @@ def test_quant_fusion(self):
+        FileCheck().check_not("aten::conv2d") \
+                   .check_not("aten::relu") \
+                   .check("quantized::conv2d_relu") \
+                   .run(m.graph_for(data))
```
Are there other checks that one can do: i.e check if there is a single inlined graph and that all weight tensors are constants?

jerryzh168 replied:
you mean the functionality of freeze? doesn't feel like that's something we should be testing here

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20354878](https://our.internmc.facebook.com/intern/diff/D20354878)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20354878](https://our.internmc.facebook.com/intern/diff/D20354878)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20354878](https://our.internmc.facebook.com/intern/diff/D20354878)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Add quantized conv2d-relu fusion pattern"


Summary:
att

Test Plan:
python test/test_jit.py

Reviewers:
mvz

Subscribers:

Tasks:

Tags:

Differential Revision: [D20354878](https://our.internmc.facebook.com/intern/diff/D20354878)

[ghstack-poisoned]
